docs/howtos/full_eval.rst

```diff
@@ -0,0 +1,197 @@
+preceding batches.
+
+
+Manual implementation
+~~~~~~~~~~~~~~~~~~~~~
+
+The last batch is manually padded to contain the same number of examples as in
+the preceding batches. The predictions for the padded examples are discarded
+from the computation.
+
+.. code-block:: python
+
+  shard = lambda x: einops.rearrange(
+      x, '(d b) ... -> d b ...', d=jax.local_device_count())
+  unshard = lambda x: einops.rearrange(x, 'd b ... -> (d b) ...')
+
+  correct = total = 0
+  for batch in ds.as_numpy_iterator():
+    images = batch['image']
+    n = len(images)
+    padding = np.zeros([per_host_batch_size - n, *images.shape[1:]], images.dtype)
+    padded_images = np.concatenate([images, padding])
+    preds = unshard(get_preds(vs_p, shard(padded_images)))[:n]
+    total += n
+    correct += (batch['label'] == preds.argmax(axis=-1)).sum()
+
+
+Using ``pad_shard_unpad()``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The above pattern, namely the pad→shard→predict→unshard→unpad sequence, can be
+extracted into a utility wrapper ``pad_shard_unpad()``, which greatly simplifies
+above evaluation loop.
+
+.. code-block:: python
+
+  correct = total = 0
+  for batch in ds.as_numpy_iterator():
+    preds = flax.jax_utils.pad_shard_unpad(get_preds)(
+        vs, batch['image'], min_device_batch=per_device_batch_size)
+    total += len(batch['image'])
+    correct += (batch['label'] == preds.argmax(axis=-1)).sum()
+
+
+Computing metrics in ``eval_step()``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Instead of returning the predictions and computing the metrics in the main
+evaluation loop, we would often want to make the metric computation part of the
+evaluation step, especially when using libraries like |jax_metrics|_, or
+|clu.metrics|_.
+
+In that case we would want to pass the metrics as a ``static_argnums`` (i.e. do
+not shard/pad it), and treat the return value as ``static_return`` too (i.e. no
+un-sharding or un-padding):
+
+.. code-block:: python
+
+  def eval_step(metrics, variables, batch):
+    print('retrigger compilation', {k: v.shape for k, v in batch.items()})
+    preds = model.apply(variables, batch['image'])
+    correct = (batch['mask'] & (batch['label'] == preds.argmax(axis=-1))).sum()
+    total = batch['mask'].sum()
+    return dict(
+        correct=metrics['correct'] + jax.lax.psum(correct, axis_name='batch'),
+        total=metrics['total'] + jax.lax.psum(total, axis_name='batch'),
+    )
+
+  eval_step = jax.pmap(eval_step, axis_name='batch')
+  eval_step = flax.jax_utils.pad_shard_unpad(
+      eval_step, static_argnums=(0, 1), static_return=True)
+
+.. |jax_metrics| replace:: ``clu.metrics``
+.. _jax_metrics: https://github.com/cgarciae/jax_metrics
+
+
+.. |clu.metrics| replace:: ``clu.metrics``
+.. _clu.metrics: https://github.com/google/CommonLoopUtils/blob/main/clu/metrics.py
+
+
+Adding "infinite padding"
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Above solution works in most cases, but it has some limitations:
```

Comment:
another solution is to let each host process indepedently and do the pmean(metrics) add the very end in a seperate pmapped program

Reply:
yes, this is actually very similar to what I do with `count_correct_p()` below

note that the above two sections `Adding "infinite padding"` and `Computing metrics in eval_step()` can easily be combined

so I think the different usecases are covered with the subsections, but if you feel there is a specific combination that should be added, feel free to add some more specific comments and I'll write it up.

Reply:
Ah indeed a parallel stoppinig criteria is not so different when you use padding. I was thinking more towards a note on what to do when you don't want to pad. You might not want to include that route for simplicity though

Reply:
but without padding different hosts would have different batch sizes? and this would also lead to re-compilations?

(in that case I think we should keep it simple with "solution=padding" since it seems superior, at the cost of a little more complicated code, but that added complication is quite small, especially in the case where one compute the metrics in the main eval loop)